src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
  */
-public final class InsertSizeMetricsCollectorSpark implements Serializable{
+public final class InsertSizeMetricsCollectorSpark implements Serializable {
     private static final long serialVersionUID = 1L;
 
-    private final InsertSizeMetrics metrics = new InsertSizeMetrics();
-    private Histogram<Integer> htslibHistogram;
+    // histograms and metrics are stored in an order that coarser level groups appear in front of finer level
+    //     (library is a level coarser than read group level)
+    //     within the same level, entries are sorted by the GroupInfo alphanumerically.
+    // the accompanying map helps retrieving the information
+    // when a particular group's name in String format is given.
+    private List<SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>>> histogramsAndMetrics = new LinkedList<>();
+    private Map<String, Integer> mapFromGroupNameToIndexInList = new HashMap<>();
+
+    // For MetricsAccumulationLevel.ALL_READS, its group name is set to the following level.
+    // This is used only in the htsjdk Histogram title.
+    private static final String ALL_READS_KEY_MACRO = "All_reads";
 
     /**
-     * Constructor and do collection work.
-     * @param filteredReads          reads that pass filters
-     * @param samRgRecords           a list of read groups
-     * @param HistogramMADTolerance  MAD tolerance when producing histogram plot
+     * Delegates actual work to utility functions.
+     *
+     * @param filteredReads         reads that pass filters
+     * @param header                header in the input
+     * @param accumLevels           accumulation level {ALL_READS, SAMPLE, LIBRARY, READ_GROUP}
+     * @param histogramMADTolerance MAD tolerance when producing histogram plot
      */
-    // TODO: * @param accumLevels            accumulation level {SAMPLE, LIBRARY, READGROUP}
     public InsertSizeMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
-                                           //final Set<MetricAccumulationLevel> accumLevels,
-                                           final List<SAMReadGroupRecord> samRgRecords,
-                                           final double HistogramMADTolerance) {
-
-        final JavaRDD<Integer> lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
-        lengthRDD.cache();
-
-        metrics.READ_PAIRS = filteredReads.count();
-
-        collectLevelInfo();
+                                           final SAMFileHeader header,
+                                           final Set<MetricAccumulationLevel> accumLevels,
+                                           final double histogramMADTolerance) {
+
+        // construct untrimmed hand rolled "histogram" (SortedMap) in three steps, because htsjdk Histogram does not play well with Serialization
+        // so first hand roll a histogram using Spark (read traversal is the bottleneck in terms of performance),
```

Comment:
I just found out about this, but Spark also includes a method for automatically computing histograms (See DoubleRDD) . Would using that help to simplify this class, or would it be too hard to have it mimic the Picard metrics tool's functionality?